### PR TITLE
Ensure track is present in charm origins when sanitising

### PR DESCRIPTION
**Pull request by jack-w-shaw**
There's a long-running bug in charmhub where if an effective channel implicitly resolves to the track 'latest', the track itself will be left out

We, however, always want a track present. Since we know a track-less channel from the charmhub api actually means the 'latest' channel, we can fill this in to ensure a track is always present

I believe the best place for this defaulting is the 'sanitiseCharmorigin' function, whose purpose is to ensure Juju can understand charmhub's weird idiosyncrasies

As a flyby, fix spelling of 'sanitize' to use proper British English 🇬🇧🇬🇧🇬🇧

## Migration and upgrade steps

It turns out corrective migrations steps already exist. This is how the bug was originally detected!
https://github.com/juju/juju/blob/3.5/state/migration_import.go#L1455-L1457

Upgrade steps, targetting 3.3.3, have been added to fill in the missing "latest" track, if none is present.

## Charmhub bug demonstration

This bug demonstrated here is the reason we need this fallback. Look to these cases:
```
$ curl -X POST -H "Content-Type: application/json" https://api.charmhub.io/v2/charms/refresh -d '
{
    "context": [],
    "actions": [
        {
            "action": "install",
            "instance-key": "507b0464-f0bf-488a-84c1-877f40610f36",
            "name": "ubuntu",
            "channel": "stable",
            "base": {"architecture":"amd64","name":"ubuntu","channel":"22.04"}
        }
    ]
}
' 2>/dev/null | jq -r '.results[0]["effective-channel"]'
stable
```
^ We should expect `latest/stable` here

We see the full `<track>/<channel>` for charms with default channels other than `latest`. See:
```
$ curl -X POST -H "Content-Type: application/json" https://api.charmhub.io/v2/charms/refresh -d '
{
    "context": [],
    "actions": [
        {
            "action": "install",
            "instance-key": "507b0464-f0bf-488a-84c1-877f40610f36",
            "name": "mysql",
            "channel": "stable",
            "base": {"architecture":"amd64","name":"ubuntu","channel":"22.04"}
        }
    ]
}
' 2>/dev/null | jq -r '.results[0]["effective-channel"]'
8.0/stable
```

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

#### Deploy some charms
```
$ juju deploy ubuntu
Deployed "ubuntu" from charm-hub charm "ubuntu", revision 24 in channel latest/stable on ubuntu@22.04/stable
$ juju deploy ubuntu ubu-edge --channel edge
Deployed "ubu-edge" from charm-hub charm "ubuntu", revision 24 in channel latest/edge on ubuntu@22.04/stable
$ juju deploy mysql
Deployed "mysql" from charm-hub charm "mysql", revision 196 in channel 8.0/stable on ubuntu@22.04/stable
```

#### Check mongo
```
$ juju mongo
> db.applications.find().pretty()
...
{
	"_id" : "5933f00e-847a-4223-8e1b-60a7a373309d:ubuntu",
	"name" : "ubuntu",
	"model-uuid" : "5933f00e-847a-4223-8e1b-60a7a373309d",
	"subordinate" : false,
	"charmurl" : "ch:amd64/focal/ubuntu-24",
	"charm-origin" : {
		"source" : "charm-hub",
		"type" : "charm",
		"id" : "DksXQKAQTZfsUmBAGanZAhpoS4dpmXel",
		"hash" : "9b2877f7ebf04b348cf40ace32fdfc6d1cc0157dea7fd6740c4ba74e0e201e5b",
		"revision" : 24,
		"channel" : {
			"track" : "latest",
			"risk" : "stable"
		},
		"platform" : {
			"architecture" : "amd64",
			"os" : "ubuntu",
			"channel" : "22.04"
		}
	},
	"charmmodifiedversion" : 0,
	"forcecharm" : false,
	"life" : 0,
	"unitcount" : 1,
	"relationcount" : 0,
	"minunits" : 0,
	"txn-revno" : NumberLong(3),
	"metric-credentials" : BinData(0,""),
	"exposed" : false,
	"scale" : 0,
	"passwordhash" : "",
	"provisioning-state" : null
}
{
	"_id" : "5933f00e-847a-4223-8e1b-60a7a373309d:ubu-edge",
	"name" : "ubu-edge",
	"model-uuid" : "5933f00e-847a-4223-8e1b-60a7a373309d",
	"subordinate" : false,
	"charmurl" : "ch:amd64/jammy/ubuntu-24",
	"charm-origin" : {
		"source" : "charm-hub",
		"type" : "charm",
		"id" : "DksXQKAQTZfsUmBAGanZAhpoS4dpmXel",
		"hash" : "9b2877f7ebf04b348cf40ace32fdfc6d1cc0157dea7fd6740c4ba74e0e201e5b",
		"revision" : 24,
		"channel" : {
			"track" : "latest",
			"risk" : "edge"
		},
		"platform" : {
			"architecture" : "amd64",
			"os" : "ubuntu",
			"channel" : "22.04"
		}
	},
	"charmmodifiedversion" : 0,
	"forcecharm" : false,
	"life" : 0,
	"unitcount" : 1,
	"relationcount" : 0,
	"minunits" : 0,
	"txn-revno" : NumberLong(3),
	"metric-credentials" : BinData(0,""),
	"exposed" : false,
	"scale" : 0,
	"passwordhash" : "",
	"provisioning-state" : null
}
{
	"_id" : "5933f00e-847a-4223-8e1b-60a7a373309d:mysql",
	"name" : "mysql",
	"model-uuid" : "5933f00e-847a-4223-8e1b-60a7a373309d",
	"subordinate" : false,
	"charmurl" : "ch:amd64/jammy/mysql-196",
	"charm-origin" : {
		"source" : "charm-hub",
		"type" : "charm",
		"id" : "F8CtvgcznoeRSupRrY2uEa67oQtBTF7x",
		"hash" : "385f6df8e9c5cc85c68924c0999f3c08787c6ba869431423d73183963a49e692",
		"revision" : 196,
		"channel" : {
			"track" : "8.0",
			"risk" : "stable"
		},
		"platform" : {
			"architecture" : "amd64",
			"os" : "ubuntu",
			"channel" : "22.04"
		}
	},
	"charmmodifiedversion" : 0,
	"forcecharm" : false,
	"life" : 0,
	"unitcount" : 1,
	"relationcount" : 2,
	"minunits" : 0,
	"txn-revno" : NumberLong(3),
	"metric-credentials" : BinData(0,""),
	"exposed" : false,
	"scale" : 0,
	"passwordhash" : "",
	"provisioning-state" : null
}
```

Notice all charms have a `track` in their `charm-origin` `channel`

### Migration

Bootstrap controllers `lxd-src` from 3.3/stable and `lxd-dst` from 3.3/edge (i.e. this PR)
```
git checkout 3.3
juju bootstrap lxd lxd-src --build-agent
```
```
git checkout JUJU-5379_always_ensure_track
juju bootstrap lxd lxd-dst --build-agent
```

Deploy ubuntu to `lxd-src` & verify
```
$ juju switch lxd-src
$ juju add-model m
$ juju deploy uibuntu
$ juju mongo
juju:PRIMARY> db.applications.find().pretty()
...
{
	"_id" : "a155d768-5eab-4315-8ade-93d8ebf00a1d:ubuntu",
	"name" : "ubuntu",
	"model-uuid" : "a155d768-5eab-4315-8ade-93d8ebf00a1d",
	"subordinate" : false,
	"charmurl" : "ch:amd64/focal/ubuntu-24",
	"charm-origin" : {
		"source" : "charm-hub",
		"type" : "charm",
		"id" : "DksXQKAQTZfsUmBAGanZAhpoS4dpmXel",
		"hash" : "9b2877f7ebf04b348cf40ace32fdfc6d1cc0157dea7fd6740c4ba74e0e201e5b",
		"revision" : 24,
		"channel" : {
			"risk" : "stable"
		},
		"platform" : {
			"architecture" : "amd64",
			"os" : "ubuntu",
			"channel" : "22.04"
		}
	},
	"charmmodifiedversion" : 0,
	"forcecharm" : false,
	"life" : 0,
	"unitcount" : 1,
	"relationcount" : 0,
	"minunits" : 0,
	"txn-revno" : NumberLong(3),
	"metric-credentials" : BinData(0,""),
	"exposed" : false,
	"scale" : 0,
	"passwordhash" : "",
	"provisioning-state" : null
}
```

Migrate & verify
```
$ Juju migrate m lxd-dst
$ juju switch lxd-dst:m
$ juju debug-log -m controller --replay | grep "origin"
Source origin for application "ubuntu" is invalid. Normalising
```
### Upgrade steps
Bootstrap a 3.3.2 controller and deploy some charms
```
git checkout v3.3.2
make install
juju bootstrap lxd lxd
juju add-model m
juju deploy ubuntu
juju deploy mysql`
```

Upgrade
```
git checkout JUJU-5379_always_ensure_track
make juju
juju upgrade-controller --build-agent
```
verify
```
$ juju mongo
juju:PRIMARY> db.applications.find().pretty()
...
{
	"_id" : "b618c215-6ed1-40a7-8181-9e38f1051336:ubuntu",
	"name" : "ubuntu",
	"model-uuid" : "b618c215-6ed1-40a7-8181-9e38f1051336",
	"subordinate" : false,
	"charmurl" : "ch:amd64/focal/ubuntu-24",
	"charm-origin" : {
		"source" : "charm-hub",
		"type" : "charm",
		"id" : "DksXQKAQTZfsUmBAGanZAhpoS4dpmXel",
		"hash" : "9b2877f7ebf04b348cf40ace32fdfc6d1cc0157dea7fd6740c4ba74e0e201e5b",
		"revision" : 24,
		"channel" : {
			"risk" : "stable",
			"track" : "latest"
		},
		"platform" : {
			"architecture" : "amd64",
			"os" : "ubuntu",
			"channel" : "22.04"
		}
	},
	"charmmodifiedversion" : 0,
	"forcecharm" : false,
	"life" : 0,
	"unitcount" : 1,
	"relationcount" : 0,
	"minunits" : 0,
	"txn-revno" : NumberLong(4),
	"metric-credentials" : BinData(0,""),
	"exposed" : false,
	"scale" : 0,
	"passwordhash" : "",
	"provisioning-state" : null
}
{
	"_id" : "b618c215-6ed1-40a7-8181-9e38f1051336:mysql",
	"name" : "mysql",
	"model-uuid" : "b618c215-6ed1-40a7-8181-9e38f1051336",
	"subordinate" : false,
	"charmurl" : "ch:amd64/jammy/mysql-196",
	"charm-origin" : {
		"source" : "charm-hub",
		"type" : "charm",
		"id" : "F8CtvgcznoeRSupRrY2uEa67oQtBTF7x",
		"hash" : "385f6df8e9c5cc85c68924c0999f3c08787c6ba869431423d73183963a49e692",
		"revision" : 196,
		"channel" : {
			"track" : "8.0",
			"risk" : "stable"
		},
		"platform" : {
			"architecture" : "amd64",
			"os" : "ubuntu",
			"channel" : "22.04"
		}
	},
	"charmmodifiedversion" : 0,
	"forcecharm" : false,
	"life" : 0,
	"unitcount" : 1,
	"relationcount" : 2,
	"minunits" : 0,
	"txn-revno" : NumberLong(3),
	"metric-credentials" : BinData(0,""),
	"exposed" : false,
	"scale" : 0,
	"passwordhash" : "",
	"provisioning-state" : null
}
```

Notice both applications have a track in their charm origin channel

### CODE REVIEW

File: core/charm/repository/sanitizer.go [hunk 1]
```diff
@@ -40,5 +40,12 @@ func sanitizeCharmOrigin(received, requested corecharm.Origin) (corecharm.Origin
 	}
 	result.Platform.OS = strings.ToLower(result.Platform.OS)
 
+	// Another problem area is the origin channel. We desire a full channel (track and risk)
+	// but if the effective channel implicitly resolves to a 'latest' channel, charmhub will
+	// just return the risk. This only happens for 'latest'. If we have no track, it must be latest
+	if result.Channel != nil && result.Channel.Track == "" {
```
Comment (nvinuesa) [inline core/charm/repository/sanitizer.go:46]: what happens if `result.Channel` is nil?

**Review by nvinuesa — APPROVED** [review]
Code is very clean and well tested, QA passes. I only have one question about the charm origin `Channel` when it's nil.

File: core/charm/repository/sanitizer.go [hunk 2]
```diff
@@ -40,5 +40,12 @@ func sanitizeCharmOrigin(received, requested corecharm.Origin) (corecharm.Origin
...
+		result.Channel.Track = "latest"
```
Comment (SimonRichardson) [inline core/charm/repository/sanitizer.go:47]: Should this not be `result.Channel.Risk = "stable"`?

File: core/charm/repository/sanitizer.go (hunk 1 above)
Comment (jack-w-shaw) [reply core/charm/repository/sanitizer.go:46]: Sometimes we have no channel, for instance, local charm origins. In this case, we should keep the channel nil

**Review by hmlanigan — CHANGES_REQUESTED** [review]
What are the QA steps for migration/upgrade from 3.3/stable to 3.3/edge? 

Why don't we need an upgrade step for 3.3/stable to 3.3/edge?

File: core/charm/repository/sanitizer.go (hunk 2 above)
Comment (hmlanigan) [reply core/charm/repository/sanitizer.go:47]: At this point, if we leave the track empty and the charm author adds a default track you could end up in an unexpected place.

File: core/charm/repository/sanitizer.go (hunk 1 above)
Comment (hmlanigan) [reply core/charm/repository/sanitizer.go:46]: Local charms should not hit this code path.

You can have a nil channel if you've requested to resolve a charm by revision. Revisions are independent from channels as the same charm revision can be in channels.

**Review by Aflynn50 — APPROVED** [review]
LGTM

File: state/migration_import.go [hunk 3]
```diff
@@ -1428,6 +1428,11 @@ func (i *importer) makeApplicationDoc(a description.Application) (*applicationDo
 // https://bugs.launchpad.net/juju/+bug/2039267
 // https://github.com/juju/python-libjuju/issues/962
 //
+// However, up until as recently as 3.4.0, tracks were often dropped from charmhub channels. This resulted
+// from a long-standing bug in the charmhub api where the track is often dropped off the effective
+// channel is it implicitly resolves to 'latest'. So if the recieved channel has no track, we can
+// be confident it should be 'latest'
+//
```
Comment (hmlanigan) [inline state/migration_import.go:1435]:
```suggestion
// Due to LP:1986547: where the track is missing from the effective channel it implicitly
// resolves to 'latest' if the charm does not have a default channel defined. So if the
// received channel has no track, we can be confident it should be 'latest'
//
```

**Review by hmlanigan — APPROVED** [review]
Some comments and tests to fix. QA went well.

File: core/charm/repository/sanitiser.go [hunk 4]
```diff
@@ -40,5 +40,12 @@ func sanitizeCharmOrigin(received, requested corecharm.Origin) (corecharm.Origin
 	}
 	result.Platform.OS = strings.ToLower(result.Platform.OS)
 
+	// Another problem area is the origin channel. We desire a full channel (track and risk)
+	// but if the effective channel implicitly resolves to a 'latest' channel, charmhub will
+	// just return the risk. This only happens for 'latest'. If we have no track, it must be latest
```
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply core/charm/repository/sanitiser.go]: Added much more context

**Pull request merged**